File: haar.rb
# encoding: utf-8

# Haarのウェーブレット解析をするためのクラス
class HaarWavelet
  # @param [Array<Float>] wave 波形
  # @param [Integer] scale_size スケールの大きさ, 2以上を指定
  def initialize(wave, scale_size)
    @wave = wave
    @difference = Array.new(scale_size)
    @difference[0] = @wave
    @weight = Array.new(scale_size)
    @weight[0] = Array.new(@wave.length, 0)
    CalcScale()
  end

  def CalcScale()
    for i in 1..@difference.length-1 do
      prev_wave = @difference[i-1]
      now_difference = Array.new(prev_wave.length >> 1)
      now_weight = Array.new(now_difference.length)

      for j in 0..now_difference.length-1 do
        weight = (prev_wave[j*2] - prev_wave[j*2+1]) * 0.5
        difference = prev_wave[j*2] - weight
        now_difference[j] = difference
        now_weight[j] = weight
      end
      @difference[i] = now_difference
      @weight[i] = now_weight
    end
  end

  # 長さを返す
  # @return [Array<Float>] スケールごとの長さ
  def Difference(scale)
    return @difference[scale]
  end

  # 重みを返す
  # @return [Array<Float>] スケールごとの重み
  def Weight(scale)
    return @weight[scale]
  end
end

File: haar.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace HaarWaveletCS
{
	/// <summary>
	/// Haarのウェーブレットを表すクラス
	/// </summary>
	class HaarWavelet
	{
		private float[] wave;
		private uint scale_size;

		private List<float[]> weight;
		private List<float[]> differernce;

		/// <summary>
		/// 入力信号をコンストラクタの段階でウェーブレット解析する
		/// </summary>
		/// <param name="wave">入力信号の配列</param>
		/// <param name="scale_size">解析するスケールの大きさ</param>
		public HaarWavelet(float[] wave, uint scale_size)
		{
			this.wave = wave;
			this.scale_size = scale_size;

			this.weight = new List<float[]>();
			this.weight.Add(new float[this.wave.Length]);

			this.differernce = new List<float[]>();
			this.differernce.Add(wave);
		}

		/// <summary>
		/// Wavelet解析をするクラス
		/// </summary>
		private void AnalyzeWavelet()
		{
			for (int i = 1; i < this.scale_size; i++)
			{
				int now_size = this.weight[i - 1].Length >> 1;
				float[] now_weight = new float[now_size];
				float[] now_difference = new float[now_size];

				for (int j = 0; j < now_size; j++)
				{
					now_weight[j] =
						(this.differernce[i - 1][j * 2] + this.differernce[i - 1][j * 2 + 1]) * 0.5f;
					now_difference[j] =
						this.differernce[i - 1][j * 2] - now_weight[j];
				}

				this.weight.Add(now_weight);
				this.differernce.Add(now_difference);
			}
		}

		/// <summary>
		/// 重みを返す
		/// </summary>
		/// <param name="scale">スケール</param>
		/// <returns>スケールに対応した重み</returns>
		public float[] Weight(int scale)
		{
			return this.weight[scale];
		}

		/// <summary>
		/// 差分を返す
		/// </summary>
		/// <param name="scale">スケール</param>
		/// <returns>スケールに対応した差分</returns>
		public float[] Difference(int scale)
		{
			return this.differernce[scale];
		}
	}
}
